### DISCUSSION TITLE
Migrer vers les Signal avec toSignal() | Bonjour Angular

### DISCUSSION THREAD

**Discussion by giscus** [bonjour-angular/apprendre#17] [General]
# Migrer vers les Signal avec toSignal() | Bonjour Angular

Migrer vers les Signal avec toSignal()

https://apprendre.bonjour-angular.com/tips/to-signal/

**Comment by manudss** [comment]
Peut être préciser dans cet article que toSignal() doit être utilisé dans un contexte d'injection. Car la méthode, doit pouvoir se désabonner de l'observable lorsque le composant est détruit (destroy). 

Les contextes d'injection ce disponible uniquement dans le constructeur, et dans les attributs de classe (tout comme la fonction inject() ). 
Cela ne fonctionnera pas si c'est en dehors de ce context : Par exemple, dans le ngOnInit. 

Si on a besoins d'y accéder, il faut utiliser la fonction runInInjectionContext:

```typescript
constructor(
    private injector: EnvironmentInjector,
  ) {}

  ngOnInit(): void {
   
    runInInjectionContext(this.injector, () => {
      this.isEmptyList = toSignal(this.selectCount().pipe(map((item) => !item)), { initialValue: true });
    });
  }
```

**Reply by KevTale** [reply]
Tu as raison ! Je ferai mieux de le préciser, c'est fait maintenant, merci pour ton feedback !

**Comment by AJ1062910** [comment]
je sais pas si c'est vraiment si simple de convertir en signal un gros projet

**Comment by AJ1062910** [comment]
mais quel est l'interet de faire ça autant faire le async dans le template

**Reply by KevTale** [reply]
L'intérêt c'est d'avoir une réactivité plus fine et plus simple grâce aux Signals.

**Reply by manudss** [reply]
@AJ1062910 Tu peux très bien continuer à utiliser async, cela fonctionne, et c'est très bien. Mais il y a des cas, ou utiliser signal, peut être bien. Par exemple, si tu doit faire un subscribe dans ton composant, qui va mettre à jour des propriétés dans ta classe. Et bien utiliser un signal permettra à ton composant, d'être au courant des changements. Sans signal, tu es obligé d'appeler le change détection manuellement. (Plus aussi, avec signal, le composant s'occupera de faire le unsubscribe. Ce qui n'est pas le cas, si tu utilises un subscribe dans ta classe.). 

Autre exemple utile, Si dans ton observable, tu as besoins d'utiliser la même valeur à plusieurs endroits dans ton code. Tu dois d'utiliser plusieurs async, mais chaque async sera une subscription différente. Ou souvent, on était obligé de mettre la valeur en paramètres dans la classe, pour y accéder à plusieurs endroits dans le html. Si tu le transformes, en signal, tu peux l'utiliser partout, dans le html, ou la classe, et tu profites de la réactivité que propose Signal. 

Donc non, il n'est pas nécessaire de tous migrer en Signal, mais dans certains cas cela peut être très utile.
